README.md
# mersenne-twister


ChangeLog.md
# Changelog for mersenne-twister

## Unreleased changes


test/Spec.hs
import qualified System.Random.MT              as MT
import           Data.List                      ( unfoldr )
import           Data.Maybe                     ( mapMaybe )
import           Data.Word                      ( Word64 )
import           System.Random                  ( RandomGen(genWord64) )
import           Text.Read                      ( readMaybe )
import           Test.Hspec

main :: IO ()
main = do
  contents <- readFile "cref/5489.txt"
  let xs  = mapMaybe readMaybe . words $ contents
      mts = word64s $ MT.fromSeed 5489
  hspec $ do
    describe "MT.hs" $ do
      it "correctly matches up with the results located in \"cref/5489.txt\""
        $ do
            and (zipWith (==) xs mts) `shouldBe` True

word64s :: RandomGen g => g -> [Word64]
word64s = unfoldr (Just . genWord64)


src/System/Random/MT.hs
module System.Random.MT
  ( MTGen
  , new
  , fromSeed
  )
where

import           Control.Arrow                  ( Arrow((&&&)) )
import           Data.Bits                      ( Bits
                                                  ( xor
                                                  , shiftL
                                                  , complement
                                                  , (.&.)
                                                  , (.|.)
                                                  , shiftR
                                                  )
                                                )
import           Data.Sequence                  ( Seq((:|>), (:<|))
                                                , (|>)
                                                , (<|)
                                                )
import qualified Data.Sequence                 as Seq
import           Data.Word                      ( Word64 )
import           Numeric                        ( readHex )
import           System.Random

-- let's define an interface

newtype MTGen = MTGen (Seq Word64)

instance RandomGen MTGen where
  genWord64 (MTGen xs) = (extract &&& MTGen) . twist $ xs
  split _ = error "i have no idea how this would be split"

instance Show MTGen where
  show (MTGen (x :<| _)) = "MTGen [" ++ show x ++ "...]"

new :: IO MTGen
new = fromSeed <$> getStdRandom uniform

fromSeed :: Word64 -> MTGen
fromSeed = MTGen . initialize

-- here's the computational stuff

-- word size (in bits)
w :: Int
w = 64

-- degree of recurrence
n :: Int
n = 312

-- middle word, an offset used to help define x (1 <= m < n)
m :: Int
m = 156

-- separation point of one word (0 <= r <= w - 1)
r :: Int
r = 31

-- coefficients of the rational form twist matrix
a :: Word64
[(a, _)] = readHex "B5026F5AA96619E9"

-- TGFSR(R) tempering bitmasks
b :: Word64
[(b, _)] = readHex "71D67FFFEDA60000"

c :: Word64
[(c, _)] = readHex "FFF7EEE000000000"

-- additional Mersenne Twister bitmask
d :: Word64
[(d, _)] = readHex "5555555555555555"

-- TGFSR(R) tempering bit shifts
s :: Int
s = 17

t :: Int
t = 37

-- additional Mersenne Twister bit shifts
u :: Int
u = 29

l :: Int
l = 43

-- an extra parameter for creating the initial state
f :: Word64
f = 6364136223846793005

-- 2 ^ (n * w - r) - 1 is a Mersenne prime

-- extract the random number that corresponds to the current state of the sequence
extract :: Seq Word64 -> Word64
extract (_ :|> x) = temper x

temper :: Word64 -> Word64
temper s0 = tempered
 where
  s1       = s0 `xor` ((s0 `shiftR` u) .&. d)
  s2       = s1 `xor` ((s1 `shiftL` s) .&. b)
  s3       = s2 `xor` ((s2 `shiftL` t) .&. c)
  tempered = s3 `xor` (s3 `shiftR` l)

-- update the sequence so that a new value can be extracted from the end of it
twist :: Seq Word64 -> Seq Word64
twist xs = Seq.drop 1 xs |> next
 where
  x0   = xs `Seq.index` 0
  x1   = xs `Seq.index` 1
  x2   = xs `Seq.index` m

  mask = (1 `shiftL` r) - 1
  applyA x = if even x then x `shiftR` 1 else (x `shiftR` 1) `xor` a

  upper = x0 .&. complement mask
  lower = x1 .&. mask
  next  = x2 `xor` applyA (upper .|. lower)

-- create a sequence of n values that are calculated based on the given seed
initialize :: Word64 -> Seq Word64
initialize = initialize' (n - 1)

initialize' :: Int -> Word64 -> Seq Word64
initialize' 0 seed = Seq.singleton seed
initialize' i seed = xs |> next
 where
  xs@(_ :|> prev) = initialize' (i - 1) seed
  next            = f * (prev `xor` (prev `shiftR` (w - 2))) + fromIntegral i


cref/mt19937-64.c
/*
   A C-program for MT19937-64 (2004/9/29 version).
   Coded by Takuji Nishimura and Makoto Matsumoto.

   This is a 64-bit version of Mersenne Twister pseudorandom number
   generator.

   Before using, initialize the state by using init_genrand64(seed)
   or init_by_array64(init_key, key_length).

   Copyright (C) 2004, Makoto Matsumoto and Takuji Nishimura,
   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:

     1. Redistributions of source code must retain the above copyright
        notice, this list of conditions and the following disclaimer.

     2. Redistributions in binary form must reproduce the above copyright
        notice, this list of conditions and the following disclaimer in the
        documentation and/or other materials provided with the distribution.

     3. The names of its contributors may not be used to endorse or promote
        products derived from this software without specific prior written
        permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER
   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

   References:
   T. Nishimura, ``Tables of 64-bit Mersenne Twisters''
     ACM Transactions on Modeling and
     Computer Simulation 10. (2000) 348--357.
   M. Matsumoto and T. Nishimura,
     ``Mersenne Twister: a 623-dimensionally equidistributed
       uniform pseudorandom number generator''
     ACM Transactions on Modeling and
     Computer Simulation 8. (Jan. 1998) 3--30.

   Any feedback is very welcome.
   http://www.math.hiroshima-u.ac.jp/~m-mat/MT/emt.html
   email: m-mat @ math.sci.hiroshima-u.ac.jp (remove spaces)
*/

#include <stdio.h>

#define NN 312
#define MM 156
#define MATRIX_A 0xB5026F5AA96619E9ULL
#define UM 0xFFFFFFFF80000000ULL /* Most significant 33 bits */
#define LM 0x7FFFFFFFULL         /* Least significant 31 bits */

/* The array for the state vector */
static unsigned long long mt[NN];
/* mti==NN+1 means mt[NN] is not initialized */
static int mti = NN + 1;

/* initializes mt[NN] with a seed */
void init_genrand64(unsigned long long seed) {
    mt[0] = seed;
    for (mti = 1; mti < NN; mti++)
        mt[mti] =
            (6364136223846793005ULL * (mt[mti - 1] ^ (mt[mti - 1] >> 62)) +
             mti);
}

/* initialize by an array with array-length */
/* init_key is the array for initializing keys */
/* key_length is its length */
void init_by_array64(unsigned long long init_key[],
                     unsigned long long key_length) {
    unsigned long long i, j, k;
    init_genrand64(19650218ULL);
    i = 1;
    j = 0;
    k = (NN > key_length ? NN : key_length);
    for (; k; k--) {
        mt[i] = (mt[i] ^
                 ((mt[i - 1] ^ (mt[i - 1] >> 62)) * 3935559000370003845ULL)) +
                init_key[j] + j; /* non linear */
        i++;
        j++;
        if (i >= NN) {
            mt[0] = mt[NN - 1];
            i = 1;
        }
        if (j >= key_length)
            j = 0;
    }
    for (k = NN - 1; k; k--) {
        mt[i] = (mt[i] ^
                 ((mt[i - 1] ^ (mt[i - 1] >> 62)) * 2862933555777941757ULL)) -
                i; /* non linear */
        i++;
        if (i >= NN) {
            mt[0] = mt[NN - 1];
            i = 1;
        }
    }

    mt[0] = 1ULL << 63; /* MSB is 1; assuring non-zero initial array */
}

/* generates a random number on [0, 2^64-1]-interval */
unsigned long long genrand64_int64(void) {
    int i;
    unsigned long long x;
    static unsigned long long mag01[2] = {0ULL, MATRIX_A};

    if (mti >= NN) { /* generate NN words at one time */

        /* if init_genrand64() has not been called, */
        /* a default initial seed is used     */
        if (mti == NN + 1)
            init_genrand64(5489ULL);

        for (i = 0; i < NN - MM; i++) {
            x = (mt[i] & UM) | (mt[i + 1] & LM);
            mt[i] = mt[i + MM] ^ (x >> 1) ^ mag01[(int)(x & 1ULL)];
        }
        for (; i < NN - 1; i++) {
            x = (mt[i] & UM) | (mt[i + 1] & LM);
            mt[i] = mt[i + (MM - NN)] ^ (x >> 1) ^ mag01[(int)(x & 1ULL)];
        }
        x = (mt[NN - 1] & UM) | (mt[0] & LM);
        mt[NN - 1] = mt[MM - 1] ^ (x >> 1) ^ mag01[(int)(x & 1ULL)];

        mti = 0;
    }

    x = mt[mti++];

    x ^= (x >> 29) & 0x5555555555555555ULL;
    x ^= (x << 17) & 0x71D67FFFEDA60000ULL;
    x ^= (x << 37) & 0xFFF7EEE000000000ULL;
    x ^= (x >> 43);

    return x;
}

/* generates a random number on [0, 2^63-1]-interval */
long long genrand64_int63(void) { return (long long)(genrand64_int64() >> 1); }

/* generates a random number on [0,1]-real-interval */
double genrand64_real1(void) {
    return (genrand64_int64() >> 11) * (1.0 / 9007199254740991.0);
}

/* generates a random number on [0,1)-real-interval */
double genrand64_real2(void) {
    return (genrand64_int64() >> 11) * (1.0 / 9007199254740992.0);
}

/* generates a random number on (0,1)-real-interval */
double genrand64_real3(void) {
    return ((genrand64_int64() >> 12) + 0.5) * (1.0 / 4503599627370496.0);
}

int main(void) {
    int i;
    // ***Modified*** this output uses the default initial seed: 5489
    for (i = 0; i < 1000; i++) {
        printf("%20llu ", genrand64_int64());
        if (i % 5 == 4)
            printf("\n");
    }
    return 0;
}
